### PR TITLE
Pass context from subscriptions through to links

### PR DESCRIPTION
**Pull request by sgtpepper43**
Resolves apollographql/apollo-feature-requests#93

Allow a context to be supplied when creating subscriptions, and pass that context through to the links.

There are a bunch of other changes in here that I'm pretty sure was made by the linter or prettier or whatever you've got going. Some pre-commit hook that automatically ran. I'll mark those changes and I can get rid of them if you want.

### Checklist:

- [x] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests

### CODE REVIEW

File: packages/apollo-client/src/core/ObservableQuery.ts [hunk 1]
```diff
@@ -63,10 +63,10 @@ export interface UpdateQueryOptions<TVariables> {
 export const hasError = (
   storeValue: QueryStoreValue,
   policy: ErrorPolicy = 'none',
-) => storeValue && (
-  storeValue.networkError ||
-  (policy === 'none' && isNonEmptyArray(storeValue.graphQLErrors))
-);
+) =>
```
Comment (sgtpepper43) [inline packages/apollo-client/src/core/ObservableQuery.ts]: Prettier changes 👇

File: packages/apollo-client/src/core/ObservableQuery.ts [hunk 2]
```diff
@@ -338,26 +337,23 @@ export class ObservableQuery<
...
     const qid = this.queryManager.generateQueryId();
 
     return this.queryManager
-      .fetchQuery(
-        qid,
-        combinedOptions,
-        FetchType.normal,
-        this.queryId,
-      )
+      .fetchQuery(qid, combinedOptions, FetchType.normal, this.queryId)
```
Comment (sgtpepper43) [inline packages/apollo-client/src/core/ObservableQuery.ts]: Prettier changes ☝️

File: packages/apollo-client/src/core/ObservableQuery.ts [hunk 3]
```diff
@@ -448,11 +445,10 @@ export class ObservableQuery<
       this.options.variables as TVariables,
       // Try to fetch the query if fetchPolicy changed from either cache-only
       // or standby to something else, or changed to network-only.
-      oldFetchPolicy !== fetchPolicy && (
-        oldFetchPolicy === 'cache-only' ||
-        oldFetchPolicy === 'standby' ||
-        fetchPolicy === 'network-only'
-      ),
+      oldFetchPolicy !== fetchPolicy &&
```
Comment (sgtpepper43) [inline packages/apollo-client/src/core/ObservableQuery.ts]: Prettier changes 👇

File: packages/apollo-client/src/core/ObservableQuery.ts [hunk 4]
```diff
@@ -608,47 +601,49 @@ export class ObservableQuery<
...
+            }
           }
-        }
-      },
-      error: onError,
-    }).catch(onError);
+        },
+        error: onError,
+      })
+      .catch(onError);
```
Comment (sgtpepper43) [inline packages/apollo-client/src/core/ObservableQuery.ts]: Prettier changes ☝️

File: packages/apollo-client/src/core/QueryManager.ts [hunk 5]
```diff
@@ -932,7 +937,7 @@ export class QueryManager<TStore> {
     options: WatchQueryOptions,
     listener: QueryListener,
   ) {
-    invariant.warn("The QueryManager.startQuery method has been deprecated");
+    invariant.warn('The QueryManager.startQuery method has been deprecated');
```
Comment (sgtpepper43) [inline packages/apollo-client/src/core/QueryManager.ts]: Prettier changes ☝️

File: packages/apollo-client/src/core/QueryManager.ts [hunk 6]
```diff
@@ -148,7 +148,7 @@ export class QueryManager<TStore> {
 
     invariant(
       !fetchPolicy || fetchPolicy === 'no-cache',
-      "fetchPolicy for mutations currently only supports the 'no-cache' policy"
+      "fetchPolicy for mutations currently only supports the 'no-cache' policy",
```
Comment (sgtpepper43) [inline packages/apollo-client/src/core/QueryManager.ts]: Prettier changes 👇

File: packages/apollo-client/src/core/QueryManager.ts [hunk 7]
```diff
@@ -948,45 +953,38 @@ export class QueryManager<TStore> {
...
-        if (!fetchPolicy || fetchPolicy !== 'no-cache') {
-          this.dataStore.markSubscriptionResult(
-            result,
-            query,
-            variables,
-          );
-          this.broadcastQueries();
-        }
+      this.getObservableFromLink<T>(query, context, variables, false).map(
+        result => {
```
Comment (sgtpepper43) [inline packages/apollo-client/src/core/QueryManager.ts]: Prettier changes 👇

File: packages/apollo-client/src/core/QueryManager.ts [hunk 8]
```diff
@@ -948,45 +953,38 @@ export class QueryManager<TStore> {
...
+          return result;
+        },
+      );
 
     if (this.transform(query).hasClientExports) {
-      const observablePromise = this.localState.addExportedVariables(
-        query,
-        variables,
-      ).then(makeObservable);
+      const observablePromise = this.localState
```
Comment (sgtpepper43) [inline packages/apollo-client/src/core/QueryManager.ts]: Prettier changes ☝️

File: packages/apollo-client/src/core/QueryManager.ts [hunk 9]
```diff
@@ -948,45 +953,38 @@ export class QueryManager<TStore> {
...
+        .addExportedVariables(query, variables, context)
+        .then(makeObservable);
```
Comment (sgtpepper43) [inline packages/apollo-client/src/core/QueryManager.ts]: Prettier changes 👇

File: packages/apollo-client/src/core/QueryManager.ts [hunk 10]
```diff
@@ -1339,11 +1340,14 @@ export class QueryManager<TStore> {
...
-    options: WatchQueryOptions;
-  }>();
+  private pollingInfoByQueryId = new Map<
+    string,
+    {
+      interval: number;
+      timeout: NodeJS.Timeout;
+      options: WatchQueryOptions;
+    }
+  >();
```
Comment (sgtpepper43) [inline packages/apollo-client/src/core/QueryManager.ts]: Prettier changes ☝️

**Comment by promontis** [comment]
@hwillson any ETA when will this be released?

**Comment by hwillson** [comment]
@promontis It will be available in the next RC release, which will probably be later today.

**Pull request merged**